File: tools/MakeAudio.java
import java.io.ByteArrayOutputStream;
import java.io.DataOutputStream;
import java.io.File;
import java.nio.file.Files;
import java.util.Random;

public final class MakeAudio {
  static final int SR = 22050;
  static final Random RNG = new Random(1945);

  public static void main(String[] args) throws Exception {
    File raw = new File(args.length > 0 ? args[0] : "app/src/main/res/raw");
    raw.mkdirs();
    writeWav(new File(raw, "sfx_shot.wav"), shot(), SR);
    writeWav(new File(raw, "sfx_boom.wav"), boom(1.15, false), 44100);
    writeWav(new File(raw, "sfx_boom_big.wav"), boom(1.75, true), 44100);
    writeWav(new File(raw, "bgm_stage.wav"), music(), SR);
    System.out.println("wrote " + raw.getAbsolutePath());
  }

  static float[] shot() {
    float dur = 0.085f;
    int n = (int) (SR * dur);
    float[] out = new float[n];
    for (int i = 0; i < n; i++) {
      float t = i / (float) SR;
      float a = (float) Math.exp(-t / 0.028) * (1f - t / dur);
      float f = 1480f - 920f * (t / dur);
      float sq = Math.sin(2 * Math.PI * f * t) >= 0 ? 1f : -1f;
      float click = noise() * (float) Math.exp(-t / 0.004) * 0.35f;
      if (t > 0.012f) click = 0f;
      out[i] = sq * a * 0.38f + click;
    }
    return out;
  }

  static float[] boom(double dur, boolean big) {
    int sr = 44100;
    int n = (int) (sr * dur);
    float[] out = new float[n];
    float lp1 = 0f;
    float lp2 = 0f;
    float hp = 0f;
    float brown = 0f;
    double phase = 0;
    double phase2 = 0;
    for (int i = 0; i < n; i++) {
      float t = i / (float) sr;
      float nse = noise();
      brown = (brown + nse) * 0.5f;
      float env = (float) Math.exp(-t / (big ? 0.48 : 0.28));
      float smoke = (float) Math.pow(Math.max(0.0, 1.0 - t / dur), 1.15);
      float punch = (float) Math.exp(-t / (big ? 0.09 : 0.06));
      float fc = (big ? 90f : 110f) + (big ? 1600f : 1200f) * (float) Math.exp(-t / (big ? 0.16 : 0.11));
      float a = 1f - (float) Math.exp(-2.0 * Math.PI * fc / sr);
      float mix = nse * 0.55f + brown * 0.45f;
      lp1 += a * (mix - lp1);
      lp2 += a * (lp1 - lp2);
      hp = mix - lp2;
      float crack = hp * punch * 0.22f;
      float dropHz = (float) ((big ? 92.0 : 108.0) * Math.exp(-t / (big ? 0.22 : 0.16)));
      phase += 2.0 * Math.PI * Math.max(48.0, dropHz) / sr;
      phase2 += 2.0 * Math.PI * Math.max(36.0, dropHz * 0.5) / sr;
      float kick = (float) (Math.sin(phase) * punch * (big ? 0.95 : 0.78));
      float sub = (float) (Math.sin(phase2) * env * (big ? 0.7 : 0.52));
      float body = lp2 * (env * 0.95f + smoke * 0.55f);
      float v = body + kick + sub + crack;
      out[i] = (float) Math.tanh(v * (big ? 1.65 : 1.4));
    }
    return peak(out, 0.94f);
  }

  static float[] music() {
    double bpm = 148;
    double beat = 60.0 / bpm;
    int bars = 8;
    int n = (int) (SR * bars * 4 * beat);
    float[] out = new float[n];
    int[] bassPat = {33, 33, 33, 33, 36, 36, 31, 31, 33, 33, 38, 38, 36, 31, 33, 33};
    int[] arpA = {57, 60, 64, 69, 64, 60, 57, 52};
    int[] arpC = {60, 64, 67, 72, 67, 64, 60, 55};
    int[] lead = {
      69, 0, 72, 0, 76, 76, 74, 0, 72, 0, 69, 71, 72, 0, 64, 0,
      67, 0, 69, 71, 72, 74, 76, 0, 74, 72, 71, 0, 69, 0, 64, 67,
      69, 0, 72, 0, 76, 79, 76, 0, 74, 0, 72, 71, 69, 0, 64, 0,
      67, 69, 71, 72, 74, 0, 72, 0, 69, 0, 64, 67, 69, 0, 57, 0
    };
    for (int i = 0; i < n; i++) {
      double t = i / (double) SR;
      double beatI = t / beat;
      int bar = ((int) (beatI / 4)) % bars;
      int eighth = ((int) (beatI * 2)) % bassPat.length;
      int sixteenth = (int) (beatI * 4);
      double posInBeat = beatI % 1.0;
      float v = 0f;
      if (((int) beatI) % 2 == 0 && posInBeat < 0.12) {
        v +=
            (float)
                (Math.sin(2 * Math.PI * (90 - 50 * posInBeat / 0.12) * t)
                    * (1 - posInBeat / 0.12)
                    * 0.42);
      }
      float hat = (sixteenth % 2 == 1) ? 1f : 0.35f;
      if ((t / (beat / 2.0)) % 1.0 < 0.06) v += noise() * hat * 0.07f;
      int midi = bassPat[eighth];
      v += sq(t * hz(midi), 0.42f) * 0.16f;
      v += (float) Math.sin(2 * Math.PI * hz(midi - 12) * t) * 0.12f;
      int[] arp = bar % 2 == 0 ? arpA : arpC;
      v += sq(t * hz(arp[sixteenth % 8]), 0.25f) * 0.07f;
      int lm = lead[((int) (beatI * 2)) % lead.length];
      if (lm != 0) {
        float gate = ((t / (beat / 2.0)) % 1.0) < 0.82 ? 1f : 0f;
        v += tri(t * hz(lm)) * 0.13f * gate;
        v += sq(t * hz(lm + 12), 0.18f) * 0.03f * gate;
      }
      out[i] = v;
    }
    int delay = (int) (0.018 * SR);
    float[] wet = out.clone();
    for (int i = delay; i < n; i++) wet[i] += out[i - delay] * 0.14f;
    return peak(wet, 0.78f);
  }

  static float hz(int midi) {
    return (float) (440.0 * Math.pow(2, (midi - 69) / 12.0));
  }

  static float sq(double phase, float width) {
    double p = phase - Math.floor(phase);
    return p < width ? 1f : -1f;
  }

  static float tri(double phase) {
    double p = phase - Math.floor(phase);
    return (float) (p < 0.5 ? 4 * p - 1 : 3 - 4 * p);
  }

  static float noise() {
    return RNG.nextFloat() * 2f - 1f;
  }

  static float[] peak(float[] in, float gain) {
    float m = 0.0001f;
    for (float s : in) m = Math.max(m, Math.abs(s));
    float[] out = new float[in.length];
    for (int i = 0; i < in.length; i++) out[i] = in[i] / m * gain;
    return out;
  }

  static void writeWav(File file, float[] samples, int rate) throws Exception {
    ByteArrayOutputStream data = new ByteArrayOutputStream();
    for (float s : samples) {
      float c = Math.max(-1f, Math.min(1f, s));
      short v = (short) (c * 32767);
      data.write(v & 0xff);
      data.write((v >> 8) & 0xff);
    }
    byte[] pcm = data.toByteArray();
    ByteArrayOutputStream bos = new ByteArrayOutputStream();
    DataOutputStream out = new DataOutputStream(bos);
    out.writeBytes("RIFF");
    writeLE(out, 36 + pcm.length);
    out.writeBytes("WAVE");
    out.writeBytes("fmt ");
    writeLE(out, 16);
    writeLE16(out, 1);
    writeLE16(out, 1);
    writeLE(out, rate);
    writeLE(out, rate * 2);
    writeLE16(out, 2);
    writeLE16(out, 16);
    out.writeBytes("data");
    writeLE(out, pcm.length);
    out.write(pcm);
    out.flush();
    Files.write(file.toPath(), bos.toByteArray());
  }

  static void writeLE(DataOutputStream out, int v) throws Exception {
    out.write(v & 0xff);
    out.write((v >> 8) & 0xff);
    out.write((v >> 16) & 0xff);
    out.write((v >> 24) & 0xff);
  }

  static void writeLE16(DataOutputStream out, int v) throws Exception {
    out.write(v & 0xff);
    out.write((v >> 8) & 0xff);
  }
}

File: tools/MakeAudio.py
#!/usr/bin/env python3
"""Procedural 16-bit PCM WAVs for Wing Strike (SFX + looping 90s shmup BGM)."""
from __future__ import annotations

import math
import os
import random
import struct
import sys
import wave

SR = 22050


def clamp(v: float) -> float:
    return -1.0 if v < -1.0 else 1.0 if v > 1.0 else v


def write_wav(path: str, samples: list[float]) -> None:
    os.makedirs(os.path.dirname(path), exist_ok=True)
    with wave.open(path, "w") as w:
        w.setnchannels(1)
        w.setsampwidth(2)
        w.setframerate(SR)
        w.writeframes(b"".join(struct.pack("<h", int(clamp(s) * 32767)) for s in samples))


def env_exp(t: float, dur: float, tau: float) -> float:
    return math.exp(-t / tau) if 0 <= t < dur else 0.0


def noise() -> float:
    return random.random() * 2.0 - 1.0


def shot(dur: float = 0.085) -> list[float]:
    n = int(SR * dur)
    out = []
    for i in range(n):
        t = i / SR
        a = env_exp(t, dur, 0.028) * (1.0 - t / dur)
        f = 1480 - 920 * (t / dur)
        sq = 1.0 if math.sin(2 * math.pi * f * t) >= 0 else -1.0
        click = noise() * env_exp(t, 0.012, 0.004) * 0.35
        out.append(sq * a * 0.38 + click)
    return out


def boom(dur: float, bass: float, body: float) -> list[float]:
    n = int(SR * dur)
    out = []
    lp = 0.0
    for i in range(n):
        t = i / SR
        nse = noise()
        lp = lp * 0.82 + nse * 0.18
        crack = nse * env_exp(t, 0.04, 0.012)
        rumble = math.sin(2 * math.pi * (78 - 40 * t / dur) * t)
        rumble += 0.45 * math.sin(2 * math.pi * (46 - 18 * t / dur) * t)
        a = env_exp(t, dur, body)
        out.append(lp * a * 0.72 + crack * 0.28 + rumble * a * bass)
    peak = max(abs(s) for s in out) or 1.0
    return [s / peak * 0.92 for s in out]


def sq(phase: float, width: float = 0.5) -> float:
    return 1.0 if (phase % 1.0) < width else -1.0


def tri(phase: float) -> float:
    p = phase % 1.0
    return 4 * p - 1 if p < 0.5 else 3 - 4 * p


def note_hz(midi: int) -> float:
    return 440.0 * (2 ** ((midi - 69) / 12.0))


def music() -> list[float]:
    bpm = 148.0
    beat = 60.0 / bpm
    bars = 8
    beats = bars * 4
    n = int(SR * beats * beat)
    out = [0.0] * n

    bass_pat = [
        33, 33, 33, 33, 36, 36, 31, 31,
        33, 33, 38, 38, 36, 31, 33, 33,
    ]
    # sixteenth arpeggios, two octaves of A minor / C
    arp_a = [57, 60, 64, 69, 64, 60, 57, 52]
    arp_c = [60, 64, 67, 72, 67, 64, 60, 55]
    lead = [
        69, 0, 72, 0, 76, 76, 74, 0,
        72, 0, 69, 71, 72, 0, 64, 0,
        67, 0, 69, 71, 72, 74, 76, 0,
        74, 72, 71, 0, 69, 0, 64, 67,
        69, 0, 72, 0, 76, 79, 76, 0,
        74, 0, 72, 71, 69, 0, 64, 0,
        67, 69, 71, 72, 74, 0, 72, 0,
        69, 0, 64, 67, 69, 0, 57, 0,
    ]

    def add(i: int, v: float) -> None:
        if 0 <= i < n:
            out[i] += v

    for i in range(n):
        t = i / SR
        beat_i = t / beat
        bar = int(beat_i // 4) % bars
        eighth = int(beat_i * 2) % len(bass_pat)
        sixteenth = int(beat_i * 4)

        pos_in_beat = (t / beat) % 1.0
        if int(beat_i) % 2 == 0 and pos_in_beat < 0.12:
            add(i, math.sin(2 * math.pi * (90 - 50 * pos_in_beat / 0.12) * t) * (1 - pos_in_beat / 0.12) * 0.42)
        hat = 1.0 if (sixteenth % 2 == 1) else 0.35
        if (t / (beat / 2.0)) % 1.0 < 0.06:
            add(i, noise() * hat * 0.07)

        midi = bass_pat[eighth]
        add(i, sq(t * note_hz(midi), 0.42) * 0.16)
        add(i, math.sin(2 * math.pi * note_hz(midi - 12) * t) * 0.12)

        arp = arp_a if bar % 2 == 0 else arp_c
        add(i, sq(t * note_hz(arp[sixteenth % 8]), 0.25) * 0.07)

        lm = lead[int(beat_i * 2) % len(lead)]
        if lm:
            gate = 1.0 if ((t / (beat / 2.0)) % 1.0) < 0.82 else 0.0
            add(i, tri(t * note_hz(lm)) * 0.13 * gate)
            add(i, sq(t * note_hz(lm + 12), 0.18) * 0.03 * gate)

    delay = int(0.018 * SR)
    wet = out[:]
    for i in range(delay, n):
        wet[i] += out[i - delay] * 0.14
    out[:] = wet

    peak = max(abs(s) for s in out) or 1.0
    return [s / peak * 0.78 for s in out]


def main() -> None:
    raw = sys.argv[1] if len(sys.argv) > 1 else os.path.join(
        os.path.dirname(__file__), "..", "app", "src", "main", "res", "raw"
    )
    raw = os.path.abspath(raw)
    write_wav(os.path.join(raw, "sfx_shot.wav"), shot())
    write_wav(os.path.join(raw, "sfx_boom.wav"), boom(0.32, bass=0.22, body=0.09))
    write_wav(os.path.join(raw, "sfx_boom_big.wav"), boom(0.55, bass=0.38, body=0.16))
    write_wav(os.path.join(raw, "bgm_stage.wav"), music())
    print("wrote", raw)


if __name__ == "__main__":
    main()
